homeassistant/components/mazda/sensor.py

```diff
@@ -1,196 +1,213 @@
 
 from . import MazdaEntity
 from .const import DATA_CLIENT, DATA_COORDINATOR, DOMAIN
 
 
-async def async_setup_entry(
-    hass: HomeAssistant,
-    config_entry: ConfigEntry,
-    async_add_entities: AddEntitiesCallback,
-) -> None:
-    """Set up the sensor platform."""
-    client = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
-    coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
-
-    entities: list[SensorEntity] = []
-
-    for index, _ in enumerate(coordinator.data):
-        entities.append(MazdaFuelRemainingSensor(client, coordinator, index))
-        entities.append(MazdaFuelDistanceSensor(client, coordinator, index))
-        entities.append(MazdaOdometerSensor(client, coordinator, index))
-        entities.append(MazdaFrontLeftTirePressureSensor(client, coordinator, index))
-        entities.append(MazdaFrontRightTirePressureSensor(client, coordinator, index))
-        entities.append(MazdaRearLeftTirePressureSensor(client, coordinator, index))
-        entities.append(MazdaRearRightTirePressureSensor(client, coordinator, index))
+@dataclass
+class MazdaSensorRequiredKeysMixin:
+    """Mixin for required keys."""
 
-    async_add_entities(entities)
+    # Suffix to be appended to the vehicle name to obtain the sensor name
+    name_suffix: str
 
+    # Function to determine the value for this sensor, given the coordinator data and the configured unit system
+    value: Callable[[dict, UnitSystem], StateType]
 
-class MazdaFuelRemainingSensor(MazdaEntity, SensorEntity):
-    """Class for the fuel remaining sensor."""
 
-    _attr_native_unit_of_measurement = PERCENTAGE
-    _attr_icon = "mdi:gas-station"
+@dataclass
+class MazdaSensorEntityDescription(
+    SensorEntityDescription, MazdaSensorRequiredKeysMixin
+):
+    """Describes a Mazda sensor entity."""
 
-    def __init__(self, client, coordinator, index) -> None:
-        """Initialize Mazda fuel remaining sensor."""
-        super().__init__(client, coordinator, index)
+    # Function to determine whether the vehicle supports this sensor, given the coordinator data
+    is_supported: Callable[[dict], bool] = lambda data: True
 
-        self._attr_name = f"{self.vehicle_name} Fuel Remaining Percentage"
-        self._attr_unique_id = f"{self.vin}_fuel_remaining_percentage"
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        return self.data["status"]["fuelRemainingPercent"]
+    # Function to determine the unit of measurement for this sensor, given the configured unit system
+    # Falls back to description.native_unit_of_measurement if it is not provided
+    unit: Callable[[UnitSystem], str | None] | None = None
 
 
-class MazdaFuelDistanceSensor(MazdaEntity, SensorEntity):
-    """Class for the fuel distance sensor."""
+def _get_distance_unit(unit_system):
+    """Return the distance unit for the given unit system."""
+    if unit_system.name == CONF_UNIT_SYSTEM_IMPERIAL:
```

Comment:
Can you explain why you are not using the automatic conversion provided by the base SensorEntity?
This should be why we moved to native_unit_of_measurement.

Reply:
If the api returns km, just set it always set it to km and let HA do the rest...

Reply:
The automatic conversion only works for temperature units. It doesn't automatically convert km to mi. See the implemention of `unit_of_measurement` in the base `SensorEntity`:
https://github.com/home-assistant/core/blob/eb50de5baa790ca4f446cfb480de44bbd277b48c/homeassistant/components/sensor/__init__.py#L346-L362